iscsi/iscsi.go

```diff
@@ -0,0 +1,349 @@
+	log.Trace.Printf("Waiting for path: %s", *devicePath)
+	if devicePath == nil {
+		return false
+	}
+
+	for i := 0; i < maxRetries; i++ {
+		var err error
+		if deviceTransport == "tcp" {
+			_, err = osStat(*devicePath)
+			if err != nil {
+				log.Trace.Printf("Stat: %s", err.Error())
+			}
+
+		} else {
+			fpath, _ := filepathGlob(*devicePath)
+			if fpath == nil {
+				err = os.ErrNotExist
+			} else {
+				// There might be a case that fpath contains multiple device paths if
+				// multiple PCI devices connect to same iscsi target. We handle this
+				// case at subsequent logic. Pick up only first path here.
+				*devicePath = fpath[0]
+			}
+		}
+		if err == nil {
+			return true
+		}
+		if !os.IsNotExist(err) {
+			log.Error.Printf("Device %s does not exist before timeout reached", *devicePath)
+			return false
+		}
+		if i == maxRetries-1 {
+			break
+		}
+		time.Sleep(time.Second * time.Duration(intervalSeconds))
+	}
+	return false
+}
+
+func getMultipathDisk(path string) (string, error) {
+	// Follow link to destination directory
+	devicePath, err := os.Readlink(path)
+	if err != nil {
+		log.Error.Printf("failed reading link: %s -- error: %s\n", path, err.Error())
+		return "", err
+	}
+	sdevice := filepath.Base(devicePath)
+	// If destination directory is already identified as a multipath device,
+	// just return its path
+	if strings.HasPrefix(sdevice, "dm-") {
+		return path, nil
+	}
+	// Fallback to iterating through all the entries under /sys/block/dm-* and
+	// check to see if any have an entry under /sys/block/dm-*/slaves matching
+	// the device the symlink was pointing at
+	dmpaths, _ := filepath.Glob("/sys/block/dm-*")
+	for _, dmpath := range dmpaths {
+		sdevices, _ := filepath.Glob(filepath.Join(dmpath, "slaves", "*"))
+		for _, spath := range sdevices {
+			s := filepath.Base(spath)
+			if sdevice == s {
+				// We've found a matching entry, return the path for the
+				// dm-* device it was found under
+				p := filepath.Join("/dev", filepath.Base(dmpath))
+				log.Trace.Printf("Found matching device: %s under dm-* device path %s", sdevice, dmpath)
+				return p, nil
+			}
+		}
+	}
+	return "", fmt.Errorf("Couldn't find dm-* path for path: %s, found non dm-* path: %s", path, devicePath)
+}
+
+// Connect attempts to connect a volume to this node using the provided Connector info
+func Connect(c Connector) (string, error) {
+
+	if c.Timeout == 0 {
+		c.Timeout = 10
+	}
+	if c.CheckInterval == 0 {
+		c.CheckInterval = 1
+	}
+
+	if c.Timeout < 0 || c.CheckInterval < 0 {
+		return "", fmt.Errorf("Invalid Timeout and CheckInterval combination, both must be positive integers. "+
+			"Timeout: %d, CheckInterval: %d", c.Timeout, c.CheckInterval)
+	}
+	var devicePaths []string
+	iFace := "default"
+	if c.Interface != "" {
+		iFace = c.Interface
+	}
+
+	// make sure our iface exists and extract the transport type
+	out, err := runCmd("iscsiadm", "-m", "iface", "-I", iFace, "-o", "show")
```

Comment:
IIUC, it's running with the CSI plugin so it's inside the container. Then I think we have a problem.

I don't think we can run multiple `iscsid`(the server `iscsiadm` connects to) on the same host. They're communicating using netlink, but seems Linux kernel iscsi doesn't support namespaced netlink. See https://github.com/rancher/longhorn/issues/112#issuecomment-405360520 .

That means, there can be only one `iscsid` running on one node. If it's installed by the OS, then no container can run it; if one container runs it, the host itself and other containers cannot run it. In the end, only one `iscsid` per node is allowed AFAIK.

Reply:
For reference, Longhorn workaround this issue by requiring users to install `iscsiadm` on the node (which came with Ubuntu server anyway) and `nsenter` to the host namespace for execution. For reference: https://github.com/rancher/longhorn-engine/blob/master/iscsi/iscsi.go

Reply:
My intent was to use the hosts iscsid, the testing I did was limited to a single container accessing it though so that might be a problem, I'll have to look at that.  There was also a method of running it in a single container (again though only one shared instance of iscsid) that I've tried in the past, I ran into some issues with it but it should be possible.

I'll check out what you did in longhorn and also get some testing to see where things are at.

Reply:
Just for reference for anybody else checking this out; the approach I used was to install iscsid on the host and then access it from the container.  I also use host network to make this work.  Here's a snippet of how I set this up.  There's likely room for improvement, I put this together quite some time ago and had success with it so never tinkered with it much.

https://gist.github.com/j-griffith/f82bd567dc963ff953adcc44d0ebd96e

Reply:
I think multiple containers with iscsiadm should work with `--net host`, though the option is not quite feasible for Longhorn.

Running it inside container maybe worse, since it may conflict with the potential `iscsid` running on the host.